### DISCUSSION TITLE
Optimize constant-sized spans of managed types via InlineArray

### DISCUSSION THREAD

**Discussion by NinoFloris** [dotnet/csharplang#8442] [Language Ideas] [answered]
It would be a nice extension of the current features to optimize the following when the size is statically known:

```cs
{ReadOnly}Span<Type> collection = [ typeof(string), typeof(int) ] // InlineArray size 2, no array alloc.
```

A span of `Type` here is a decently motivating example but ideally this would work for any managed type. Today this introduces an array alloc.

Initially I considered whether expanding allowed `stackalloc` types might be relevant here. However as it still defaults to T* it feels unwise to expand it to (likely unpinned) managed types.

**Comment by jnm2** [comment, marked as answer]
Great suggestion! The spec already allows optimizations such as this; it's at the discretion of the compiler. There's an umbrella issue for compiler optimizations at: https://github.com/dotnet/roslyn/issues/68785

**Reply by NinoFloris** [reply]
Thanks, it seems it's already implemented indeed... I tested this in a multi-targeted project where InlineArray understandably isn't used on the older TFMs. Rider's IL view then was not showing me the output for the selected 8.0 TFM but for one of the older ones, my bad!
